crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs

```diff
@@ -2883,61 +2922,195 @@ impl Analyzer<'_, '_> {
+                                        cause: vec![],
+                                    }
+                                    .into());
+                                }
+                            }
+                        }
+                        Type::Union(ty) => {
+                            if ty.types.iter().any(|v| v.is_any()) {
+                                return Ok(());
+                            }
+                            // type A = Uppercase<`aB${string | number}`> - valid
+                            // type A = Uppercase<`aB${string | { x: string }}`>; - invalid
+                            let is_valid_union = ty.types.iter().all(|v| {
+                                v.is_num_like()
+                                    || v.is_bigint_like()
+                                    || v.is_bool_like()
+                                    || v.is_str_like()
+                                    || v.is_null()
+                                    || v.is_undefined()
+                            });
+
+                            if !is_valid_union {
+                                return Err(ErrorKind::AssignFailed {
+                                    span: ty.span(),
+                                    left: box Type::StringMapping(to.clone()),
+                                    right_ident: None,
+                                    right: box r.clone(),
+                                    cause: vec![],
+                                }
+                                .into());
+                            }
+                        }
+                        _ => {
+                            if !(v.is_any()
+                                || v.is_num_like()
+                                || v.is_bigint_like()
+                                || v.is_bool_like()
+                                || v.is_str_like()
+                                || v.is_null()
+                                || v.is_undefined())
+                            {
+                                return Err(ErrorKind::AssignFailed {
+                                    span: to.span(),
+                                    left: box Type::StringMapping(to.clone()),
+                                    right_ident: None,
+                                    right: box r.clone(),
+                                    cause: vec![],
+                                }
+                                .into());
+                            }
                         }
                     }
+                }
 
-                    return Err(ErrorKind::AssignFailed {
-                        span: r.span(),
-                        left: box Type::StringMapping(to.clone()),
-                        right_ident: None,
-                        right: box r.clone(),
-                        cause: vec![],
+                return Ok(());
+            }
+            Type::Alias(alias_ty) => {
```

Comment:
I think this branch should be `unreachable!()`

Reply:
You're right, I had assumed it was for this case
```ts
type B = Lowercase<string>;
type C = Lowercase<B>;
```
but here B is `ref type` which I means I need to re-add the `type::ref` match arm :skull:

Reply:
Actually, we need to it because the ref type will be converted to an alias type. What do you think about this?

```rs
            Type::Alias(alias_ty) => {
                return self.assign_to_intrinsic(data, to, &alias_ty.ty, opts);
            }
            Type::Ref(ref_ty) => {
                if let RTsEntityName::Ident(r_ident) = &ref_ty.type_name {
                    if let Ok(Some(ty)) = self.find_type(&r_ident.into()) {
                        let ty = ty.into_iter().map(|v| v.into_owned()).collect::<Vec<Type>>();
                        return self.assign_to_intrinsic(data, to, &ty.clone()[0], opts);
                    }
                }
            }
```

if we keep alias unreachable the below fails
```ts
type B = Lowercase<string>;
type C = Lowercase<B>;
```

`B` is first `ref` then gets converted to `alias` in in `self.find_type`

Reply:
Removed `Type::Alias` from this match and added a `Type::Ref` match arm in `expand_intrinsic_types` function.